src/chains/filecoin/filecoin/src/database.ts

```diff
@@ -1,7 +1,173 @@
+    let db: LevelUp;
+    if (store) {
+      this.#rootStore = encode(store, levelupOptions);
+      // @ts-ignore - I get an error in ts-node that I don't get running ttsc:
+      // Argument of type 'AbstractLevelDOWN<any, any>' is not assignable to parameter of type 'string'
+      // But this error just doesn't make any sense. I'm just going to ignore the error
+      db = levelup(this.#rootStore!, {});
+    } else {
+      let directory = this.#options.dbPath;
+      if (!directory) {
+        const dirInfo = await dir(tmpOptions);
+        directory = dirInfo.path;
+        this.#cleanupDirectory = dirInfo.cleanup;
+
+        // don't continue if we closed while we were waiting for the dir
+        if (this.#closed) return this.#cleanup();
+      }
+      this.directory = directory;
+
+      // specify an empty `prefix` for browser-based leveldown (level-js)
+      const leveldownOpts = { prefix: "" };
+      const store = encode(leveldown(directory, leveldownOpts), levelupOptions);
+      this.#rootStore = store;
+      db = levelup(store, {});
+    }
+
+    // don't continue if we closed while we were waiting for the db
+    if (this.#closed) return this.#cleanup();
+
+    const open = db.open();
+
+    this.db = db;
+    await open;
+
+    // don't continue if we closed while we were waiting for it to open
+    if (this.#closed) return this.#cleanup();
+
+    this.tipsets = sub(db, "t", levelupOptions);
+    this.blocks = sub(db, "b", levelupOptions);
+
+    this.#initialized = true;
+    return this.emit("ready");
+  };
+
+  /**
+   * Call `batch` to batch `put` and `del` operations within the same
+   * event loop tick of the provided function. All db operations within the
+   * batch _must_ be executed synchronously.
+   * @param fn Within this function's event loop tick, all `put` and
+   * `del` database operations are applied in a single atomic operation. This
+   * provides a single write call and if any individual put/del's fail the
+   * entire operation fails and no modifications are made.
+   * @returns a Promise that resolves to the return value
+   * of the provided function.
+   */
+  public batch<T>(fn: () => T) {
+    if (!this.initialized) {
+      throw new Error(
+        "Could not call Database.batch as the database isn't initialized yet."
+      );
+    }
+
+    const rootDb = this.#rootStore!.db;
+    const batch = this.db!.batch();
+
+    const originalPut = rootDb.put;
+    const originalDel = rootDb.del;
+
+    rootDb.put = batch.put.bind(batch);
+    rootDb.del = batch.del.bind(batch);
+    let prom: Promise<T>;
+    try {
+      const ret = fn();
+      // PSA: don't let vscode (or yourself) rewrite this to `await` the
+      // `batch.write` call. The `finally` block needs to run _before_ the
+      // write promise has resolved.
+      prom = batch.write().then(() => ret);
+    } finally {
+      rootDb.put = originalPut;
+      rootDb.del = originalDel;
+    }
+    return prom;
```

Comment:
I'm curious if you reviewed this batch implementation of mine at all. It let's us do nice things and all, but I do worry that I missed something.

Reply:
I didn't, but I can! I kind of took the database stuff for granted

Reply:
👌 This implementation looks sane to me. I think the only thing I'd add was a test to ensure the atomicity. I see that this is used in `blockchain.ts::#saveNewBlock` and works (aka tests work), but I do think a targeted set of tests would be good to verify this logic works as it looks. However, I don't see why it wouldn't work!

Reply:
Issue opened for that. Thanks! https://github.com/trufflesuite/ganache-core/issues/873